File: LeetCode/93复原IP地址.go
func restoreIpAddresses(s string) []string {
	var ans []string
	var ip []string
	backtracking(s, 0, &ans, ip)
	return ans
}

func backtracking(s string, ind int, ans *[]string, ip []string) {
	if ind == len(s) && len(ip) == 4 {
		temp := ip[0] + "." + ip[1] + "." + ip[2] + "." + ip[3]
		*ans = append(*ans, temp)
		return
	}
	for i := ind; i < len(s); i++ {
		ip = append(ip, s[ind:i+1])
		// 表示当前的一个字段长度属于正常范畴,且ip字段少于4,且该字段是正常
		if i-ind+1 <= 3 && len(ip) <= 4 && isValid(s, ind, i) {
			backtracking(s, i+1, ans, ip)
		} else {
			return
		}
		ip = ip[:len(ip)-1]
	}
	return
}

func isValid(s string, left int, right int) bool {
	num, _ := strconv.Atoi(s[left : right+1])
	if right-left+1 > 1 && s[left] == '0' {
		return false
	}
	if num > 255 {
		return false
	}
	return true
}

File: LeetCode/93复原IP地址.java
class Solution {
    List<String> ans = new ArrayList<>();
    public List<String> restoreIpAddresses(String s) {
        if (s.length() > 12) return ans;
        backtracking(s, 0, 0);
        return ans;
    }

    public void backtracking(String s, int ind, int cnt) {
        if (cnt == 3) {
            if (isValid(s, ind, s.length() - 1)) {
                ans.add(s);
            }
            return ;
        }
        for (int i = ind; i < s.length(); i++) {
            if (isValid(s, ind, i)) {
                s = s.substring(0, i + 1) + "." + s.substring(i + 1);
                cnt++;
                backtracking(s, i + 2, cnt);
                cnt--;
                s = s.substring(0, i + 1) + s.substring(i + 2);
            } else break;
        }
    }

    public boolean isValid(String s, int left, int right) {
        if (left > right) return false;
        if (s.charAt(left) == '0' && left != right) return false;
        int num = 0;
        for (int i = left; i <= right; i++) {
            if (s.charAt(i) > '9' || s.charAt(i) < '0') return false;
            num = num * 10 + s.charAt(i) - '0';
            if (num > 255) return false;
        }
        return true;
    }
}